source/mir/combinatorics.d

```diff
@@ -0,0 +1,734 @@
+    k = non negative integer, k <= n
+
+Returns:
+    Binomial coefficient
+*/
+
+Result binomial(Result, NType, KType)(NType n, KType k)
+{
+    Result result = 1;
+    if (k > n)
+    {
+        result = 0;
+        return result;
+    }
+    if (k > n - k)
+    {
+        k = n - k;
+    }
+    // make a copy of n (could be a custom type)
+    for (KType i = 1, m = n; i <= k; i++, m--)
+    {
+        // check whether an overflow can happen
+        // hasMember!(Result, "max") doesn't work with dmd2.068 and ldc 0.17
+        static if (is(typeof((inout int = 0)
+        {
+            if (Result.max) {}
+        })))
+        {
+            if (result / i > Result.max / m) return 0;
+            result = result / i * m + result % i * m / i;
+        }
+        else
+        {
+            result = result * m / i;
+        }
+    }
+    return result;
+}
+
+/// ditto
+size_t binomial(size_t n, size_t k) pure nothrow @safe @nogc
+{
+    return binomial!size_t(n, k);
+}
+
+///
+unittest
+{
+    assert(binomial(5, 2) == 10);
+    assert(binomial(6, 4) == 15);
+    assert(binomial(3, 1) == 3);
+
+    import std.bigint;
+    assert(binomial!BigInt(1000, 10) == BigInt("263409560461970212832400"));
+}
+
+pure nothrow @safe @nogc unittest
+{
+    assert(binomial(5, 1) == 5);
+    assert(binomial(5, 0) == 1);
+    assert(binomial(1, 2) == 0);
+    assert(binomial(1, 0) == 1);
+    assert(binomial(1, 1) == 1);
+    assert(binomial(2, 1) == 2);
+    assert(binomial(2, 1) == 2);
+}
+
+unittest
+{
+    // test larger numbers
+    assert(binomial(100, 10) == 17310309456440);
+    assert(binomial(999, 5) == 8209039793949);
+    assert(binomial(300, 10) == 1398320233241701770LU);
+
+    // test overflow
+    assert(binomial(500, 10) == 0);
+
+    // all parameters as custom types
+    import std.bigint;
+    BigInt n = 1010, k = 9;
+    assert(binomial!BigInt(n, k) == BigInt("2908077120956865974260"));
+}
+
+/**
+Lazily computes all _permutations of `r` using $(WEB
+en.wikipedia.org/wiki/Heap%27s_algorithm, Heap's algorithm).
+
+While generating a new item is in `O(k)` (amortized `O(1)`),
+the number of permutations is `|r|!`.
+
+Returns:
+A forward range the elements of which are an $(XREF range,
+indexed) view into `r`.
+*/
+auto permutations(Range)(Range r)
+if (isRandomAccessRange!Range && hasLength!Range)
+{
+    static struct Permutations
+    {
+        size_t[] indices, state;
+        Range r;
+
+        this(Range r)
+        {
+            import std.range : iota;
+            import std.array : array;
+
+            this.r = r;
+            state = r.length ? new size_t[r.length-1] : null;
+            indices = iota(size_t(r.length)).array;
+            _empty = r.length == 0;
+        }
+
+        private bool _empty;
+
+        @property auto front()
+        {
+            import std.range : indexed;
+            return r.indexed(indices);
+        }
+
+        void popFront()
+        {
+            void next(int n)
```

Comment:
Better options: looks like this is just simple loop `for(size_t h = 0; state[n] != h; state[h]++) { ... }` where `return` replaced with `break` and `h` is `n-2` (h would be a little bit faster).